pybossa/task_creator_helper.py

```diff
@@ -40,19 +40,22 @@ def s3_conn_type():
     return current_app.config.get('S3_CONN_TYPE')
 
 
-def get_task_expiration(current_expiration):
+def get_task_expiration(current_expiration, create_time=None):
     """
     Find the appropriate expiration to be added to a task with expiring data.
     If no expiration is set, return the data expiration; otherwise, return
     the smallest between the current expiration and the data expiration.
     current_expiration can be a iso datetime string or a datetime object
     """
     validity = current_app.config.get('TASK_EXPIRATION', 60)
-    return _get_task_expiration(current_expiration, validity)
+    return _get_task_expiration(current_expiration, create_time, validity)
 
 
-def _get_task_expiration(current_expiration, validity):
-    task_exp = get_now_plus_delta_ts(days=validity)
+def _get_task_expiration(current_expiration, create_time, validity):
+    if create_time is None:
+        task_exp = get_now_plus_delta_ts(days=validity)
+    else:
+        task_exp = get_time_plus_delta_ts(create_time, days=validity)
```

Comment:
1. Task create_time can/should never be null. in such case I think code should throw exception and not compute expiration from current time. Because, with this, user would still be able to set expiration beyond max allowed expiration which needs to be avoided.
2. Though it seems get_now_... function not needed in this scenario, get_now_... & get_time_.. can be refactored easily into a more generic single function as get_time_... passing current_time when delta needed from current time.

Reply:
Addressing point 1:
Task `create_time` null check is handled here: https://github.com/bloomberg/pybossa/pull/836/files#diff-8964fd9572e087e46289906ef7ec80e1e5eb472d1d9caef28c8972a2453fd656L82. We already handle the case where create_time is null before entering `get_task_expiration` function

In `_get_task_expiration`, the only way `create_time` is null is if we intentionally pass it in ourselves.

As for point 2, refactored via https://github.com/bloomberg/pybossa/pull/836/commits/48d172ad76c431e72526cf85bd4db82a8f88be83

Reply:
> In `get_task_expiration`, the only way `create_time` is null is if we intentionally pass it in ourselves.

Actually, it cannot be passed and is to be set by the system. Check the code [here](https://github.com/bloomberg/pybossa/blob/main/pybossa/api/task.py#L54), as `created` is reserved field, it cannot be passed as null or any other value.